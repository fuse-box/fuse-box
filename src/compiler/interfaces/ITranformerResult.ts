import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
