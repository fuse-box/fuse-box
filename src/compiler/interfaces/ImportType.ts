export enum ImportType {
  REQUIRE = 1,
  FROM = 2,
  RAW_IMPORT = 3,
  DYNAMIC = 4,
}
