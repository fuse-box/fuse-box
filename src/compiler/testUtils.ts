import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
