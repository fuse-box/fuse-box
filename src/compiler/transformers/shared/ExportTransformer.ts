import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import {
  ES_MODULE_EXPRESSION,
  createExports,
  createVariableDeclaration,
  isDefinedLocally,
} from '../../Visitor/helpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';
import { GlobalContext } from '../../program/GlobalContext';

const IGNORED_DECLARATIONS = {
  [ASTType.InterfaceDeclaration]: 1,
  [ASTType.TypeAliasDeclaration]: 1,
};
function considerDecorators(node: ASTNode) {
  let statement: ASTNode = node.declaration;
  if (
    node.declaration.type === 'ClassDeclaration' &&
    node.declaration.decorators &&
    node.declaration.decorators.length
  ) {
    statement = createVariableDeclaration(node.declaration.id.name, node.declaration);
    // assigne decorators to the statement
    statement.decorators = node.declaration.decorators;
    statement.$fuse_decoratorForClassIdentifier = node.declaration.id.name;
    // remove them from the original declaration (in order for the decorators to be instered after)
    node.declaration.decorators = [];
  }
  return statement;
}
const INTERESTED_NODES = {
  ExportAllDeclaration: 1,
  ExportDefaultDeclaration: 1,
  ExportNamedDeclaration: 1,
};

export function ExportTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      const definedLocallyProcessed: Record<string, number> = {};
      const compilerOptions = props.compilerOptions;
      let injectEsModuleStatement = compilerOptions.esModuleStatement === true;
      let esModuleStatementInjected = !injectEsModuleStatement;
      return {
        onEachNode: (visit: IVisit) => {
          const global = visit.globalContext as GlobalContext;

          if (global.exportAfterDeclaration) {
            const node = visit.node;
            const definedLocally = isDefinedLocally(node);

            if (definedLocally && !visit.node.$fuse_visited) {
              const newNodes = [];

              const response: IVisitorMod = {};

              for (const localVar of definedLocally) {
                // we don't want to add it multiple times
                if (definedLocallyProcessed[localVar.name]) break;
                definedLocallyProcessed[localVar.name] = 1;
                const targetAfter = global.exportAfterDeclaration[localVar.name];
                if (targetAfter) {
                  if (!response.appendToBody) response.appendToBody = [];
                  for (const item of targetAfter.targets) {
                    const ast = createExports(global.namespace, item, {
                      name: localVar.name,
                      type: 'Identifier',
                    });
                    response.appendToBody.push(ast);
                  }
                }
              }
              if (newNodes.length) {
                node.$fuse_visited = true;

                // console.log('new nodes', newNodes);
                // return { replaceWith: [node].concat(newNodes) };
                response.insertAfterThisNode = newNodes;
              }
              if (response.insertAfterThisNode || response.appendToBody) {
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
            }
          }
        },
        onTopLevelTraverse: (visit: IVisit) => {
          const node = visit.node;

          const global = visit.globalContext as GlobalContext;

          const type = node.type;
          if (!INTERESTED_NODES[type]) return;

          // remove export interface
          // export interface HelloWorld{}
          if (type === 'ExportNamedDeclaration') {
            if (node.declaration && IGNORED_DECLARATIONS[node.declaration.type]) {
              return { removeNode: true };
            }
          }

          /**
           * **************************************************************************
           * Export with source
           * export { one, two } from "./source"
           *
           * Needs to be simply replaced, no tracking involved
           */
          if (node.source) {
            const sourceVariable = global.getModuleName(node.source.value);

            // export * from "a"
            if (node.type === 'ExportAllDeclaration') {
              const reqExpression: ASTNode = {
                arguments: [
                  {
                    type: 'Literal',
                    value: node.source.value,
                  },
                ],
                callee: {
                  name: 'require',
                  type: 'Identifier',
                },
                type: 'CallExpression',
              };
              if (props.onRequireCallExpression) {
                props.onRequireCallExpression(ImportType.FROM, reqExpression);
              }
              const response: IVisitorMod = {
                replaceWith: {
                  arguments: [
                    {
                      name: 'exports',
                      type: 'Identifier',
                    },
                    reqExpression,
                  ],
                  callee: {
                    computed: false,
                    object: {
                      name: 'Object',
                      type: 'Identifier',
                    },
                    property: {
                      name: 'assign',
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  },
                  type: 'CallExpression',
                },
              };

              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }

            // export { foo } from "./bar"
            // creating
            //    var obj = require("module")
            const exportedNodes: Array<ASTNode> = [
              {
                declarations: [
                  {
                    id: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    init: {
                      arguments: [
                        {
                          type: 'Literal',
                          value: node.source.value,
                        },
                      ],
                      callee: {
                        name: 'require',
                        type: 'Identifier',
                      },
                      type: 'CallExpression',
                    },
                    type: 'VariableDeclarator',
                  },
                ],
                kind: 'var',
                type: 'VariableDeclaration',
              },
            ];

            if (type === 'ExportNamedDeclaration') {
              for (const specifier of node.specifiers) {
                exportedNodes.push(
                  createExports(global.namespace, specifier.exported.name, {
                    computed: false,
                    object: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    property: {
                      name: specifier.local.name,
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  }),
                );
              }
            }
            const response: IVisitorMod = { replaceWith: exportedNodes };
            // make sure exports.__eModule is injected
            if (!esModuleStatementInjected) {
              esModuleStatementInjected = true;
              response.prependToBody = [ES_MODULE_EXPRESSION];
            }
            return response;
          }

          /**
           * Handling export default declaratio
           * *********************************************************
           *    export default func
           */
          if (type === 'ExportDefaultDeclaration') {
            if (node.declaration) {
              if (node.declaration.type === ASTType.InterfaceDeclaration) {
                return { removeNode: true };
              }
              // Looking at this case:
              //      console.log(foo)
              //      export default function foo(){}
              if (node.declaration.type === 'FunctionDeclaration' || node.declaration.type === 'ClassDeclaration') {
                if (!node.declaration.id) {
                  node.declaration.id = { name: '__DefaultExport__', type: 'Identifier' };
                }

                // if there are decorators on the class we need to transform and assigne to a variable
                /**
                 * @dec
                 * export default class A {
                 *
                 * }
                 *
                 * Should be converted to:
                 *
                 * let A = class A {
                 * }
                 */
                let statement: ASTNode = considerDecorators(node);
                const response: IVisitorMod = {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, 'default', {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
              const response: IVisitorMod = {
                replaceWith: createExports(global.namespace, 'default', node.declaration),
              };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
          }

          if (type === 'ExportNamedDeclaration') {
            /**
             * ******************************************************************************
             * CASE with variable exports
             *
             *      export { name1, name2 }
             *
             * Will be "soft" replaced if those defined earlier
             * Then if it's not defined we need to wait for those (they might come later)
             */

            const newNodes = [];

            if (node.specifiers && node.specifiers.length) {
              for (const specifier of node.specifiers) {
                // if the variable present in the scope...
                // or picked up by the identifierReplacement ( we can leave that for GlobalContextTranformer)
                if (
                  (visit.scope && visit.scope.locals[specifier.local.name]) ||
                  global.identifierReplacement[specifier.local.name]
                ) {
                  newNodes.push(
                    createExports(global.namespace, specifier.exported.name, {
                      name: specifier.local.name,
                      type: 'Identifier',
                    }),
                  );
                } else {
                  // if none was decleared, it must be declared earlier (so we check if later with onEachNode in this transformer )
                  let exportAfterDeclaration = global.exportAfterDeclaration;
                  if (!exportAfterDeclaration) {
                    global.exportAfterDeclaration = exportAfterDeclaration = {};
                  }

                  if (!exportAfterDeclaration[specifier.local.name]) {
                    exportAfterDeclaration[specifier.local.name] = { targets: [] };
                  }
                  // it can have multiple exports of the same variable. For example
                  // export {foo, foo as foo1}

                  exportAfterDeclaration[specifier.local.name].targets.push(specifier.exported.name);
                }
              }
              const response: IVisitorMod = { replaceWith: newNodes };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
            /**
             * ************************************************************************
             * Export objects directly
             *
             *  export function hello(){}
             *  export class Bar(){}
             */
            if (node.declaration) {
              if (node.declaration.id && node.declaration.id.name) {
                const statement = considerDecorators(node);
                return {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, node.declaration.id.name, {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
              }
              if (node.declaration.type === 'VariableDeclaration') {
                if (node.declaration.declare) {
                  return { ignoreChildren: true, removeNode: true };
                }
                /**
                 * ******************************************************************
                 * Exports defined constants
                 *
                 *    export const foo = 1, bar = 3;
                 */
                if (node.declaration.declarations) {
                  let newNodes = [];
                  for (const declaration of node.declaration.declarations) {
                    if (!declaration.declare) {
                      global.identifierReplacement[declaration.id.name] = {
                        first: global.namespace,
                        second: declaration.id.name,
                      };

                      if (declaration.init) {
                        // check if there is INIT
                        // we night have something like this:
                        //    export var FooBar;
                        newNodes.push(createExports(global.namespace, declaration.id.name, declaration.init));
                      }
                    }
                  }
                  const response: IVisitorMod = { replaceWith: newNodes };

                  // make sure exports.__eModule is injected
                  if (!esModuleStatementInjected) {
                    esModuleStatementInjected = true;
                    response.prependToBody = [ES_MODULE_EXPRESSION];
                  }
                  return response;
                }
              }
            }
          }
        },
      };
    },
  };
}
