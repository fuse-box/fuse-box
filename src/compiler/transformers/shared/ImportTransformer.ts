import * as path from 'path';
import { GlobalContext } from '../../program/GlobalContext';
import { ITransformer } from '../../program/transpileModule';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { ImportType } from '../../interfaces/ImportType';

// FIX SHORTHANDS
export function ImportTransformer(options?: ITransformerSharedOptions): ITransformer {
  options = options || {};
  return {
    onTopLevelTraverse: (visit: IVisit) => {
      const node = visit.node;
      const global = visit.globalContext as GlobalContext;

      if (node.type === 'ImportDeclaration') {
        // converts "./foo/bar.hello.js" to foo_bar_hello_js_1 (1:1 like typescript does)
        const variable = path.basename(node.source.value).replace(/\.|-/, '_') + '_' + global.getNextIndex();

        node.specifiers.forEach(specifier => {
          if (specifier.type === 'ImportSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: specifier.imported.name,
            };
          } else if (specifier.type === 'ImportDefaultSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: 'default',
            };
          } else if (specifier.type === 'ImportNamespaceSpecifier') {
            // only if we have more than one specifier
            // for instance
            // i=mport MySuperClass, * as everything from "module-name";

            // in every other case like:
            // import * as everything from "module-name";
            // we don't need to do anything, since the variables should match

            global.identifierReplacement[specifier.local.name] = {
              first: variable,
            };

            //if (node.specifiers.length > 1) afterStatement = defineVariable(variable, specifier.local.name);
          }
        });

        return {
          onComplete: () => {
            const reqStatement = createRequireStatement(node.source.value, node.specifiers.length && variable);
            // when everything is finished we need to check if those variables have been used at all
            // they were all unused we need remove the require/import statement at all

            // assuming we have no specififers and this import HAS side effects
            // e.g import "./module"
            const parent = visit.parent;
            const property = visit.property;
            if (node.specifiers.length === 0) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.RAW_IMPORT, reqStatement.reqStatement);
                }
                parent[property].splice(index, 1, reqStatement.statement);
              }
              return;
            }

            let atLeastOneInUse = false;
            for (const specifier of node.specifiers) {
              const traced = global.identifierReplacement[specifier.local.name];
              if (traced && traced.inUse) {
                atLeastOneInUse = true;
                break; // we just need to know if we need to keep the node
              }
            }
            // doing a manual replace
            if (atLeastOneInUse) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                parent[property].splice(index, 1, reqStatement.statement);
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.FROM, reqStatement.reqStatement);
                }
              }
            } else {
              const index = parent[property].indexOf(node);
              if (index > -1) parent[property].splice(index, 1);
            }
          },
        };
      }
    },
  };
}
