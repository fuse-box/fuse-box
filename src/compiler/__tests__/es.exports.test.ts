import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { ImportType } from '../interfaces/ImportType';
import { initCommonTransform } from '../testUtils';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { CommonTSfeaturesTransformer } from '../transformers/ts/CommonTSfeaturesTransformer';
import { EnumTransformer } from '../transformers/ts/EnumTransformer';

const testTranspile = (props: { code: string; compilerOptions?: ICompilerOptions; jsx?: boolean }) => {
  return initCommonTransform({
    code: props.code,
    compilerOptions: props.compilerOptions || {},
    jsx: props.jsx,
    transformers: [ImportTransformer(), EnumTransformer(), ExportTransformer(), CommonTSfeaturesTransformer()],
  });
};

describe('Es exports tests', () => {
  describe('export typescript objects', () => {
    it('should remove interface', () => {
      const result = testTranspile({
        code: `
         export interface Hey{}
         alert(1)
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should remove export default interface', () => {
      const result = testTranspile({
        code: `
         export default interface Hey{}
         alert(1)
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should remove type', () => {
      const result = testTranspile({
        code: `
         export type Some = Map<string,string>;
         alert(2)
          `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Object export / function/ class', () => {
    it('should export function', () => {
      const result = testTranspile({
        code: `
         export function hello(){}
          `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Mix import and export', () => {
    it('should export object with keys', () => {
      const result = testTranspile({
        code: `
        import stuff from "./stuff";
        export { stuff  };
            `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys (import all)', () => {
      const result = testTranspile({
        code: `
        import * as stuff from "./stuff";
        export { stuff  };
            `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys (import all) with alias', () => {
      const result = testTranspile({
        code: `
        import * as stuff from "./stuff";
        export { stuff as oi  };
            `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Export var with value', () => {
    it('should export a variable 1', () => {
      const result = testTranspile({
        code: `
        import { Observable } from '../Observable';
        import { noop } from '../util/noop';
        export var NEVER = new Observable(noop);
        console.log(NEVER);
            `,
      });
      expect(result.code).toMatchSnapshot();
    });
    it('should trace down an undefined variable', () => {
      const result = testTranspile({
        code: `
        export var FooBar;
        (function (FooBar) {
        })(FooBar || (FooBar = {}));
            `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should trace down all varaiables ignore assignements (local scope)', () => {
      const result = testTranspile({
        code: `
        export var __assign = function() {
          __assign = Object.assign || function __assign() {};
          return __assign.apply(this, arguments);
        };
            `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export a variable 1', () => {
      const result = testTranspile({
        code: `
        export const a = function() {
          console.log(Foo);
        };
        console.log(a);
        export var Foo = { foo: "bar" };
            `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should respect locals within an anon function', () => {
      const result = testTranspile({
        code: `
        import TweenLite, { _gsScope, globals } from './TweenLite.js';

        some(function() {
          var TweenMax = {};
          TweenMax.version = "2.0.2";
          TweenMax.hey = "1";
          console.log(TweenMax);
          return TweenMax;
        });
        export var Foo = globals.TweenMax;
            `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Object export', () => {
    it('should export object with keys', () => {
      const result = testTranspile({
        code: `
          const name1 = 2;
          const name2 = 3;
          export {name1, name2}
            `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export enum', () => {
      const result = testTranspile({
        code: `
        enum Foo {
          Oi,
        }
        export { Foo };
            `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys2 (late) 1', () => {
      const result = testTranspile({
        code: `
          export {name1 as fun, name2}
          const name1 = 1;
          const name2 = 2;
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export the same object with different names', () => {
      const result = testTranspile({
        code: `
          export {foo, foo as foo1}
          const foo = 1;

          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export the same object with different names (multiple refs in the code)', () => {
      const result = testTranspile({
        code: `
          var AnimationDriver = /** @class */ (function() {
            function AnimationDriver() {}
            AnimationDriver.NOOP = new NoopAnimationDriver();
            return AnimationDriver;
          })();
          export { AnimationDriver, AnimationDriver as ÉµAnimationDriver };

          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys2 (late)  function', () => {
      const result = testTranspile({
        code: `
          export {name1 as fun, name2}
          function name1(){}
          const name2 = 2;
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys2 (late) function reversed order', () => {
      const result = testTranspile({
        code: `
          function name1(){}
          const name2 = 2;
          export {name1 as fun, name2}
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys2 (late) 2 ( function )', () => {
      const result = testTranspile({
        code: `
          console.log(1);
          export {name1 as fun, name2}
          function name1() {}
          console.log(2);
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export object with keys2 (late) 2 ( class )', () => {
      const result = testTranspile({
        code: `
          console.log(1);
          export {MySuperClass as fun, name2}
          class MySuperClass {}
          console.log(2);
          `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Exports const', () => {
    it('should export constants', () => {
      const result = testTranspile({
        code: `
      export const foo = 1, bar = 3
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export constants and use globally', () => {
      const result = testTranspile({
        code: `
      export const foo = 1, bar = 3;
      function test(){
        console.log(foo)
      }
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export constants and avoid collision', () => {
      const result = testTranspile({
        code: `
      export const foo = 1, bar = 3;
      function test(foo){
        console.log(foo)
      }
          `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Export default declaration', () => {
    it('should export default ', () => {
      const result = testTranspile({
        code: `
        const add = function(){}
        export default add;
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export default (should play nicely with import) ', () => {
      const result = testTranspile({
        code: `
        import add from "./some";
        export default add;
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export default (inverse order) ', () => {
      const result = testTranspile({
        code: `
        export default add;
        function add(){}
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should not replace the scope ', () => {
      const result = testTranspile({
        code: `
        function hello() {
          console.log(add);
        }
        export default add;
        function add() {}
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export expression ', () => {
      const result = testTranspile({
        code: `
        export default { add : add }
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export a shorthand expression with import reference ', () => {
      const result = testTranspile({
        code: `
        import hey from "some_package";
        export default { hey }
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export default function with a name', () => {
      const result = testTranspile({
        code: `
        console.log(foo)
        export default function foo(){}
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should export obj', () => {
      const result = testTranspile({
        code: `
        import * as core from '@uirouter/core';
        export { core };
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export default function without a name', () => {
      const result = testTranspile({
        code: `
        export default function(){}
          `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export default class without a name', () => {
      const result = testTranspile({
        code: `
        export default class {}
          `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Export from source', () => {
    it('should handle shorhand', () => {
      const result = testTranspile({
        code: `
        import zipWith from './zipWith.js';
        const foo = 1;
        export default {
          zipWith, foo
        }
          `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export 1', () => {
      const result = testTranspile({
        code: `export { name } from "./foo" `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export 1 with alias', () => {
      const result = testTranspile({
        code: `export { name as hey } from "./foo" `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export 2 vars', () => {
      const result = testTranspile({
        code: `export { oi, name as hey } from "./foo" `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export default', () => {
      const result = testTranspile({
        code: `export { default } from "./foo" `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should export all', () => {
      const result = testTranspile({
        code: `export * from "./foo" `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should emit require call', () => {
      const result = testTranspile({
        code: `export * from "./foo" `,
      });
      expect(result.requireStatementCollection).toEqual([
        {
          importType: ImportType.FROM,
          statement: {
            arguments: [{ type: 'Literal', value: './foo' }],
            callee: { name: 'require', type: 'Identifier' },
            type: 'CallExpression',
          },
        },
      ]);
    });
  });

  describe('Default interop', () => {
    it('should export with interop', () => {
      const result = testTranspile({
        code: `export {default} from "./foo"`,
        compilerOptions: { esModuleInterop: true },
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should export with interop', () => {
      const result = testTranspile({
        code: `export {default as oi} from "./foo"`,
        compilerOptions: { esModuleInterop: true },
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Edge cases', () => {
    describe('Overriden compiler reserved exports', () => {
      it('User defines a compiler reserved keywors "exports"', () => {
        const result = testTranspile({
          code: `
            var exports = {};
            export { exports as default };
            exports.foo = function(){}
          `,
        });

        expect(result.code).toMatchSnapshot();
      });
    });
  });

  describe('export type', () => {
    it('should ignore export type"', () => {
      const result = testTranspile({
        code: `
          export type { foo } from "bar"
          console.log(1)
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should remove empty export declaration"', () => {
      const result = testTranspile({
        code: `
          export {}
        `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });
});
