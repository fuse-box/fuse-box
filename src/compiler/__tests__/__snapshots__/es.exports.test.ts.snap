// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Es exports tests Export default declaration Should export obj 1`] = `
"var core_1 = require(\\"@uirouter/core\\");
exports.core = core_1;
"
`;

exports[`Es exports tests Export default declaration should export a shorthand expression with import reference  1`] = `
"var some_package_1 = require(\\"some_package\\");
exports.default = {
  hey: some_package_1.default
};
"
`;

exports[`Es exports tests Export default declaration should export default  1`] = `
"const add = function () {};
exports.default = add;
"
`;

exports[`Es exports tests Export default declaration should export default (inverse order)  1`] = `
"exports.default = add;
function add() {}
"
`;

exports[`Es exports tests Export default declaration should export default (should play nicely with import)  1`] = `
"var some_1 = require(\\"./some\\");
exports.default = some_1.default;
"
`;

exports[`Es exports tests Export default declaration should export default class without a name 1`] = `
"class __DefaultExport__ {}
exports.default = __DefaultExport__;
"
`;

exports[`Es exports tests Export default declaration should export default function with a name 1`] = `
"console.log(foo);
function foo() {}
exports.default = foo;
"
`;

exports[`Es exports tests Export default declaration should export default function without a name 1`] = `
"function __DefaultExport__() {}
exports.default = __DefaultExport__;
"
`;

exports[`Es exports tests Export default declaration should export expression  1`] = `
"exports.default = {
  add: add
};
"
`;

exports[`Es exports tests Export default declaration should not replace the scope  1`] = `
"function hello() {
  console.log(add);
}
exports.default = add;
function add() {}
"
`;

exports[`Es exports tests Export from source should export 1 1`] = `
"var foo_1 = require(\\"./foo\\");
exports.name = foo_1.name;
"
`;

exports[`Es exports tests Export from source should export 1 with alias 1`] = `
"var foo_1 = require(\\"./foo\\");
exports.hey = foo_1.name;
"
`;

exports[`Es exports tests Export from source should export 2 vars 1`] = `
"var foo_1 = require(\\"./foo\\");
exports.oi = foo_1.oi;
exports.hey = foo_1.name;
"
`;

exports[`Es exports tests Export from source should export all 1`] = `
"Object.assign(exports, require(\\"./foo\\"))
"
`;

exports[`Es exports tests Export from source should export default 1`] = `
"var foo_1 = require(\\"./foo\\");
exports.default = foo_1.default;
"
`;

exports[`Es exports tests Export from source should handle shorhand 1`] = `
"var zipWith_js_1 = require(\\"./zipWith.js\\");
const foo = 1;
exports.default = {
  zipWith: zipWith_js_1.default,
  foo
};
"
`;

exports[`Es exports tests Export var with value should export a variable 1 1`] = `
"var Observable_1 = require(\\"../Observable\\");
var noop_2 = require(\\"../util/noop\\");
exports.NEVER = new Observable_1.Observable(noop_2.noop);
console.log(exports.NEVER);
"
`;

exports[`Es exports tests Export var with value should export a variable 1 2`] = `
"exports.a = function () {
  console.log(exports.Foo);
};
console.log(exports.a);
exports.Foo = {
  foo: \\"bar\\"
};
"
`;

exports[`Es exports tests Export var with value should respect locals within an anon function 1`] = `
"var TweenLite_js_1 = require(\\"./TweenLite.js\\");
some(function () {
  var TweenMax = {};
  TweenMax.version = \\"2.0.2\\";
  TweenMax.hey = \\"1\\";
  console.log(TweenMax);
  return TweenMax;
});
exports.Foo = TweenLite_js_1.globals.TweenMax;
"
`;

exports[`Es exports tests Export var with value should trace down an undefined variable 1`] = `
"(function (FooBar) {})(exports.FooBar || (exports.FooBar = {}));
"
`;

exports[`Es exports tests Exports const should export constants 1`] = `
"exports.foo = 1;
exports.bar = 3;
"
`;

exports[`Es exports tests Exports const should export constants and avoid collision 1`] = `
"exports.foo = 1;
exports.bar = 3;
function test(foo) {
  console.log(foo);
}
"
`;

exports[`Es exports tests Exports const should export constants and use globally 1`] = `
"exports.foo = 1;
exports.bar = 3;
function test() {
  console.log(exports.foo);
}
"
`;

exports[`Es exports tests Mix import and export should export object with keys (import all) 1`] = `
"var stuff_1 = require(\\"./stuff\\");
exports.stuff = stuff_1;
"
`;

exports[`Es exports tests Mix import and export should export object with keys (import all) with alias 1`] = `
"var stuff_1 = require(\\"./stuff\\");
exports.oi = stuff_1;
"
`;

exports[`Es exports tests Mix import and export should export object with keys 1`] = `
"var stuff_1 = require(\\"./stuff\\");
exports.stuff = stuff_1.default;
"
`;

exports[`Es exports tests Object export / function/ class should export function 1`] = `
"function hello() {}
exports.hello = hello;
"
`;

exports[`Es exports tests Object export should export object with keys 1`] = `
"const name1 = 2;
const name2 = 3;
exports.name1 = name1;
exports.name2 = name2;
"
`;

exports[`Es exports tests Object export should export object with keys2 (late)  function 1`] = `
"function name1() {}
const name2 = 2;
exports.fun = name1;
exports.name2 = name2;
"
`;

exports[`Es exports tests Object export should export object with keys2 (late) 1 1`] = `
"const name1 = 1;
const name2 = 2;
exports.fun = name1;
exports.name2 = name2;
"
`;

exports[`Es exports tests Object export should export object with keys2 (late) 2 ( class ) 1`] = `
"console.log(1);
class MySuperClass {}
console.log(2);
exports.fun = MySuperClass;
"
`;

exports[`Es exports tests Object export should export object with keys2 (late) 2 ( function ) 1`] = `
"console.log(1);
function name1() {}
console.log(2);
exports.fun = name1;
"
`;

exports[`Es exports tests Object export should export object with keys2 (late) function reversed order 1`] = `
"function name1() {}
const name2 = 2;
exports.fun = name1;
exports.name2 = name2;
"
`;

exports[`Es exports tests export typescript objects should remove interface 1`] = `
"alert(1);
"
`;

exports[`Es exports tests export typescript objects should remove type 1`] = `
"alert(2);
"
`;
